src/main/java/graphql/util/ParallelTraverser.java

```diff
@@ -0,0 +1,172 @@
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountedCompleter;
+import java.util.concurrent.ForkJoinPool;
+import java.util.function.Function;
+
+import static graphql.Assert.assertNotNull;
+import static graphql.Assert.assertTrue;
+import static graphql.util.TraversalControl.ABORT;
+import static graphql.util.TraversalControl.CONTINUE;
+import static graphql.util.TraversalControl.QUIT;
+
+@Internal
+public class ParallelTraverser<T> {
+
+    private final Function<? super T, Map<String, ? extends List<T>>> getChildren;
+    private final Map<Class<?>, Object> rootVars = new ConcurrentHashMap<>();
+
+
+    private Object sharedContextData;
+
+
+    private ParallelTraverser(Function<? super T, Map<String, ? extends List<T>>> getChildren, Object sharedContextData) {
+        this.getChildren = assertNotNull(getChildren);
+        this.sharedContextData = sharedContextData;
+    }
+
+    public static <T> ParallelTraverser<T> parallelTraverser(Function<? super T, ? extends List<T>> getChildren) {
+        return parallelTraverser(getChildren, null);
+    }
+
+    public static <T> ParallelTraverser<T> parallelTraverser(Function<? super T, ? extends List<T>> getChildren, Object sharedContextData) {
+        return new ParallelTraverser<>(wrapListFunction(getChildren), sharedContextData);
+    }
+
+    public static <T> ParallelTraverser<T> parallelTraverserWithNamedChildren(Function<? super T, Map<String, ? extends List<T>>> getNamedChildren, Object sharedContextData) {
+        return new ParallelTraverser<>(getNamedChildren, sharedContextData);
+    }
+
+
+    private static <T> Function<? super T, Map<String, ? extends List<T>>> wrapListFunction(Function<? super T, ? extends List<T>> listFn) {
+        return node -> {
+            List<T> childs = listFn.apply(node);
+            return Collections.singletonMap(null, childs);
+        };
+    }
+
+    public ParallelTraverser<T> rootVars(Map<Class<?>, Object> rootVars) {
+        this.rootVars.putAll(assertNotNull(rootVars));
+        return this;
+    }
+
+    public ParallelTraverser<T> rootVar(Class<?> key, Object value) {
+        rootVars.put(key, value);
+        return this;
+    }
+
+    public void traverse(T root, TraverserVisitor<? super T> visitor) {
+        traverse(Collections.singleton(root), visitor);
+    }
+
+    public void traverse(Collection<? extends T> roots, TraverserVisitor<? super T> visitor) {
+        traverseImpl(roots, visitor);
+    }
+
+    public DefaultTraverserContext<T> newRootContext(Map<Class<?>, Object> vars) {
+        return newContextImpl(null, null, vars, null, true);
+    }
+
+
+    public void traverseImpl(Collection<? extends T> roots, TraverserVisitor<? super T> visitor) {
+        assertNotNull(roots);
+        assertNotNull(visitor);
+
+        DefaultTraverserContext<T> rootContext = newRootContext(rootVars);
+        ForkJoinPool.commonPool().invoke(new CountedCompleter<Void>() {
```

Comment:
I think the ForkJoinPool to use MUST be a parameter.  Sure we can default it to commonPool() but as is this is now making a threading decision on behalf of the consuming code.

Lets allow people to specify it.  How many threads in a pool is ALWAYS an application concern, not a library concern

Reply:
yes .. I changed. ParallelTraverser  is marked as Internal for now btw.